libraries/AP_HAL/utility/RingBuffer.cpp

```diff
@@ -121,20 +106,81 @@ uint32_t ByteBuffer::peekbytes(uint8_t *data, uint32_t len)
         return 0;
     }
     uint32_t n;
-    const uint8_t *b = readptr(n);
+    auto b = readptr(n);
     if (n > len) {
         n = len;
     }
 
-    // perform first memcpy
-    memcpy(data, b, n);
-    data += n;
+    iovec[0].data = const_cast<uint8_t *>(b);
+    iovec[0].len = n;
+
+    if (len > n) {
```

Comment:
nitpick, but I'd say to invert the condition so it's clear you are returning 0, 1 or  2. So here you'd do:
```c++
if (len <= n) {
    return 1;
}
```

and then do the rest below